docs/docs/contributing/debugging.md

```diff
@@ -0,0 +1,332 @@
+  at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
+  at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
+  at scala.collection.immutable.List.foreach(List.scala:392)
+  at dotty.tools.dotc.typer.Namer.index(Namer.scala:722)
+  at dotty.tools.dotc.typer.Namer.recur$1(Namer.scala:484)
+  at dotty.tools.dotc.typer.Namer.indexExpanded(Namer.scala:501)
+  at dotty.tools.dotc.typer.Namer.index(Namer.scala:474)
+  at dotty.tools.dotc.typer.FrontEnd.enterSyms$$anonfun$1(FrontEnd.scala:69)
+  at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
+  at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:41)
+  at dotty.tools.dotc.typer.FrontEnd.enterSyms(FrontEnd.scala:71)
+  at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:100)
+  at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:158)
+  at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
+  at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
+  at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
+  at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
+  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
+  at dotty.tools.dotc.Run.runPhases$5(Run.scala:170)
+  at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:178)
+  at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
+  at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:102)
+  at dotty.tools.dotc.Run.compileUnits(Run.scala:185)
+  at dotty.tools.dotc.Run.compileSources(Run.scala:120)
+  at dotty.tools.dotc.Run.compile(Run.scala:104)
+  at dotty.tools.dotc.Driver.doCompile(Driver.scala:34)
+  at dotty.tools.dotc.Driver.process(Driver.scala:172)
+  at dotty.tools.dotc.Driver.process(Driver.scala:141)
+  at dotty.tools.dotc.Driver.process(Driver.scala:153)
+  at dotty.tools.dotc.Driver.main(Driver.scala:180)
+  at dotty.tools.dotc.Main.main(Main.scala)
+```
+
+So that tree was created at:
+
+```
+  at dotty.tools.dotc.ast.desugar$.moduleDef(Desugar.scala:804)
+```
+
+Since all the stack frames above it are technical frames executing the tree creation command, and the frame in question is the location where the intent of the tree creation was expressed.
+
+### Via tracer
+Some objects may not be `Positioned` and hence their creation site is not debuggable via the technique in the section above. Say you target a tree at `Typer`'s `typed` method as follows:
+
+```scala
+if (tree.show == """println("Hello World")""") {
+  val creationSite = "<creation site stack here>"
+  println(creationSite)
+}
+```
+
+In other words, you have a reference to the object and want to know were it was created. To do so, go to the class definition of that object. In our case, `tree` is a [`Tree`](https://github.com/lampepfl/dotty/blob/10526a7d0aa8910729b6036ee51942e05b71abf6/compiler/src/dotty/tools/dotc/ast/Trees.scala#L52). Now, create a new `val` member of that type:
```

Comment:
It's a bit weird to use this technique on a tree, when trees already have `-Ydebug-tree-with-id`, maybe use a different type for the example ?

Reply:
True. I've logged it at lampepfl/dotty-knowledge#6, will address with other doc issues later.